hypothesis-python/src/hypothesis/ghostwriter.py

```diff
@@ -0,0 +1,192 @@
+    # `from_type()` so that `get_type_hints()` can resolve any forward
+    # references for us.
+    if param.annotation is not inspect.Parameter.empty:
+        return infer
+    # If there's no annotation and no default value, the user will have to
+    # fill this in later.  We use nothing() as a placeholder to this effect.
+    if param.default is inspect.Parameter.empty:
+        return st.nothing()
+    # If our default value is an Enum or a boolean, we assume that any value
+    # of that type is acceptable.  Otherwise, we only generate the default.
+    if isinstance(param.default, bool):
+        return st.booleans()
+    if isinstance(param.default, enum.Enum):
+        return st.sampled_from(type(param.default))
+    return st.just(param.default)
+
+
+def _get_params(func: Callable) -> Dict[str, inspect.Parameter]:
+    var_param_kinds = (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)
+    params = list(inspect.signature(func).parameters.values())
+    return OrderedDict((p.name, p) for p in params if p.kind not in var_param_kinds)
+
+
+def _get_strategies(*funcs: Callable) -> Dict[str, st.SearchStrategy]:
+    given_strategies = {}  # type: Dict[str, st.SearchStrategy]
+    for _, f in enumerate(funcs):
+        params = _get_params(f)
+        builder_args = {k: _strategy_for(v) for k, v in params.items()}
+        strat = st.builds(f, **builder_args).wrapped_strategy  # type: ignore
+
+        args, kwargs = strat.mapped_strategy.wrapped_strategy.element_strategies
+        if args.element_strategies:
+            raise NotImplementedError("Expected to pass everything as kwargs")
+
+        for k, v in zip(kwargs.keys, kwargs.mapped_strategy.element_strategies):
+            given_strategies[k] = v
+
+    indexer = list(_get_params(f)).index
+    return OrderedDict(sorted(given_strategies.items(), key=lambda kv: indexer(kv[0])))
+
+
+def _make_test(
+    *funcs: Callable, ghost: str, test_body: str, except_: Tuple[Type[Exception], ...]
+) -> str:
+    # the common elements of each ghostwriter
+    given_strategies = _get_strategies(*funcs)
+    given_args = ", ".join(
+        "{}={!r}".format(k, v)
+        .replace("sampled_from([False, True])", "booleans()")
+        .replace("just(None)", "none()")
+        .replace(".filter(_can_hash)", "")  # hide the special case for Decimal('snan')
+        for k, v in given_strategies.items()
+    )
+    for name in st.__all__:
+        given_args = given_args.replace("{}(".format(name), "st.{}(".format(name))
+    imports = {f.__module__ for f in funcs}
+    if "typing." in given_args:  # pragma: no cover
+        # After improving the strategy-unwrapping logic, I no longer have
+        # examples where from_type(typing.XYZ) appears in reprs...
+        # but nor am I willing to bet that this will never occur for user code
+        # when the worst-case is a useless import vs an import error.
+        imports.add("typing")
```

Comment:
Should we maybe add a comment above this asking the user to report it to us if this happens?

Reply:
Thinking this through, taking out this speculative logic will at worst result in a bug report, so I've just taken it out entirely.